design_notes/XDN03 - xlang Type System.md

```diff
@@ -15,3 +15,142 @@ status: draft
 This design note describes the xlang's [type system](https://en.wikipedia.org/wiki/Type_system).
 The type system is designed to interoperate with a variety of object-oriented programming
 languages in a natural and familiar way.
+
+## Namespaces
+
+A namespace is a naming scope used to organize code and avoid naming collisions. All type
+categories in the xlang type system (enums, structs, delegates, interfaces, and runtime classes)
+except for fundamental types live in a namespace.
+
+Namespaces may contain other namespaces.
+
+xlang namespaces and type names are case preserving but insensitive. This means that you cannot
+have namespaces or type names that vary only by case. For example: you cannot have `Foo.SomeType`
+and `foo.SomeType` nor can you have `Foo.SomeType` and `Foo.someType`. However, you can have
+`Foo.SomeType` and `foo.AnotherType`. In this case `SomeType` and `AnotherType` would be considered
+as belonging to the same namespace.
+
+## Fundamental Types
+
+The xlang Type system includes a core set of built-in primitive types.
+
+xlang Type | Type Description
+---------- | ----------------------------------------------------------
+Int16      | a 16 bit signed integer
+Int32      | a 32 bit signed integer
+Int64      | a 64 bit signed integer
+UInt8      | an 8 bit unsigned integer
+UInt16     | a 16 bit unsigned integer
+UInt32     | a 32 bit unsigned integer
+UInt64     | a 64 bit unsigned integer
+Single     | a 32-bit IEEE 754 floating point number
+Double     | a 64-bit IEEE 754 floating point number
+Char16     | a 16-bit non-numeric value representing a UTF-16 code unit
+Boolean    | an 8-bit Boolean value
+String     | an immutable sequence of Char16s used to represent text
+Guid       | A 128-bit standard [universally unique identifier](https://en.wikipedia.org/wiki/Universally_unique_identifier)
+Object     | An xlang object of unknown type
+
+> Note: string encoding for xlang is currently under review. In particular, it has not been decided
+if xlang will support UTF-8 strings in some capacity. A design note will be commissioned for this
+decision.
+
+## Enums
+
+An enum type is a distinct value type with a set of named values. Each named value in an enum
+corresponds to a constant integer value.
+
+Each enum type has an underlying integral type. The only legal underlying integral types for enums
+in xlang are Int32 and UInt32.
+
+An enum type with an underlying integral type of UInt32 is considered a Flags enum. Flags enums
```

Comment:
@jonwis : So I can't have a non-flags enum that is also UInt32 for making compat with existing code easier?  Can I specify the underlying type? What if I specify the underlying type as int32 but mark it with the flags attribute?

@kennykerr : There have also been requests for 64-bit enums and unsigned enums that aren't flags (addresses).

Reply:
@jonwis midlrt enforces the constraint that non-flags enums are int32, flags enums are unit32

Reply:
FYI, I've added "open issues" sections to several parts of this doc. I've added expanding the types usable as enums. Assuming these issues are still open when I push this PR to master, I'll open issues to track them (and add their IDs to the doc)